---
title: "NHLBI-186 ACAT1 in Human CD4 T cells Bulk RNA-seq "
subtitle: "Session 3: Over Representation Analysis for 8C, 8A, 11C, and 11A"
date: "`r Sys.Date()`"
author: Yue (Gary) Zhang, PhD
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango
    code_folding: hide
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=TRUE}
#.libPaths('/Users/zhangy68/R/R_4.4.2')
knitr::opts_chunk$set(warning = T, message = T,echo = F) 
```

```{r Initiate, results='hide'}
## initiate ---------------------------------------------------------
packages_to_load = c('dplyr','ggplot2','gplots','limma','stringr','usedist',
                     'cowplot','GGally','scales','rstatix','wesanderson', 'readxl',
                     'Biobase','reshape','edgeR','plotly','DT','tidyr','ComplexHeatmap','ggfun','ggrepel', 'maSigPro','DESeq2',"org.Hs.eg.db","AnnotationDbi","clusterProfiler","fgsea")
# Load packages without printing output
invisible(suppressPackageStartupMessages({
    lapply(packages_to_load, library, character.only = TRUE)
}))

# set variables 
setwd('./')
set.seed('123')
col21 <- rev(c("tomato1","darkblue","turquoise1","lightblue","darkred","mediumblue","purple","bisque",
               "greenyellow","yellow","violetred2","darkgreen","darkgoldenrod1","deeppink3","cadetblue4",
               "orchid2","seagreen3","purple4","dodgerblue2","red","gray27"))

```

```{r function section}
source('utils.R')
```

```{r load data}
load("../data/Session1.DGE_analysis_deseq2_2samples_p3p4.RData")
```

```{r prep gene list}
# Splitting the geneID column by "|" into geneEnsembl and geneName columns
DEG <- stat
# Remove the rows with duplicated gene names
dupGene = DEG$geneName[duplicated(DEG$geneName)]
#no duplicated genes

```

# Volcano plot

```{r volcano plot}
df <- DEG

# Ensure df has the necessary columns
if (!all(c("logFC", "P.Value") %in% colnames(df))) {
  stop("Dataframe must contain 'logFC' and 'P.Value' columns")
}

# Compute -log10 of the P-value
df$negLogP <- -log10(df$P.Value)

# Define significance thresholds
logFC_threshold <- 0
pval_threshold <- 0.05

df$Significance <- "Not Significant"
df$Significance[df$logFC > logFC_threshold & df$P.Value < pval_threshold] <- "Upregulated"
df$Significance[df$logFC < -logFC_threshold & df$P.Value < pval_threshold] <- "Downregulated"

# Count the number of upregulated and downregulated genes
num_upregulated <- sum(df$Significance == "Upregulated")
num_downregulated <- sum(df$Significance == "Downregulated")

# Create the volcano plot
volcano_plot <- ggplot(df, aes(x = logFC, y = negLogP, color = Significance)) +
  geom_point(alpha = 0.6) +
  scale_color_manual(values = c("Upregulated" = "red", "Downregulated" = "blue", "Not Significant" = "gray")) +
  theme_minimal() +
  geom_vline(xintercept = c(-logFC_threshold, logFC_threshold), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(pval_threshold), linetype = "dashed", color = "black") +
  labs(title = "Volcano Plot",
       subtitle = paste("Downregulated:", num_downregulated, "genes; Upregulated:", num_upregulated, "genes"),
       x = "Log Fold Change",
       y = "-Log10 P-value",
       color = "Significance (based on P value and LogFC)")

# Print the plot
print(volcano_plot)

```

# Over-representation Analysis 

```{r cl fun}
#msigbr to download all genes and their GO IDs #Ref:http://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html
msigDB <-  msigdbr::msigdbr(species = "Homo sapiens")

```

Over-representation (or enrichment) analysis (ORA) is a statistical method that determines whether genes from pre-defined gene sets (ex: those belongs to a specific GO term or KEGG pathway) are present more than expected (over-represented) in the list of differentially expressed genes. We use clusterProfiler (https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) to perform ORA. 

Here we select genes with p value < 0.05 as differentially expressed genes, and separate them into up and down regulated gene lists using 0 as the logFC cutoff. For each gene list, we iterate ORA with six pre-defined gene sets, including Hallmark, KEGG, Reactome, GOBP, VAX and IMMUNESIGDB. Pathways having adjusted p value < 0.05 are reported.

## Upregulation {.tabset}

__Selecting genes with logFC > 0 & P.Value < 0.05; Showing gene sets with adj P val < 0.05.__

```{r run enrich }
DEG.sig = subset(DEG, logFC > 0 & P.Value < 0.05)
create_dt(DEG.sig,caption = paste0(nrow(DEG.sig), ' genes selected for ORA analysis'))
DEG.sig = DEG.sig %>% pull(geneID)

pval_cutoff = 0.05
qval_cutoff = 0.05

REACTOME.out = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "REACTOME",    pval = pval_cutoff, qval = qval_cutoff)  
KEGG.out     = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "KEGG",        pval = pval_cutoff, qval = qval_cutoff)
HALLMARK.out = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "HALLMARK",    pval = pval_cutoff, qval = qval_cutoff) 
GO.out       = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "GO",          pval = pval_cutoff, qval = qval_cutoff) 
VAX.out      = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "VAX",         pval = pval_cutoff, qval = qval_cutoff)  
IMMUN.out    = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "IMMUNESIGDB", pval = pval_cutoff, qval = qval_cutoff) 
```

### Summary of significant pathways 

```{r show enricher 2}
enricher.summary = data.frame(Type = c('Hallmark','KEGG','Reactome','GO','VAX','IMMUNESIGDB'),
                              Number_of_significant_sets = c(nrow(HALLMARK.out),nrow(KEGG.out),nrow(REACTOME.out),nrow(GO.out),nrow(VAX.out),nrow(IMMUN.out)))
create_dt(enricher.summary, paste0('Significant gene sets (p.adj < ', qval_cutoff, ")"))

```

### Hallmarks

```{r show enr hall}
tmp = HALLMARK.out
create_dt(tmp,'Significant hallmarks gene sets')
```

### KEGG

```{r show enr keg}
tmp = KEGG.out
create_dt(tmp,'Significant KEGG gene sets')
```

### Reactome

```{r show enr rea}
tmp = REACTOME.out
create_dt(tmp,'Significant reactome gene sets')
```

### GO BP

```{r show enr go,fig.height=10}
tmp = GO.out
create_dt(tmp,'Significant GO BP gene sets')
```

### VAX

```{r show enr vax,fig.height=10}
tmp = VAX.out
create_dt(tmp,'Significant VAX gene sets')
```

### IMMUNESIGDB

```{r show enr imm,fig.height=10}
tmp = IMMUN.out
create_dt(tmp,'Significant IMMUNESIGDB gene sets')
```

## Down-regulation {.tabset}

__Selecting genes with logFC < 0 & P.Value < 0.05; Showing gene sets with adj P val < 0.05.__

```{r run enrich 2}
DEG.sig = subset(DEG, logFC < 0 & P.Value < 0.05)
create_dt(DEG.sig,caption = paste0(nrow(DEG.sig), ' genes selected for ORA analysis'))
DEG.sig = DEG.sig %>% pull(geneID)

pval_cutoff = 0.05
qval_cutoff = 0.05

REACTOME.out = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "REACTOME",    pval = pval_cutoff, qval = qval_cutoff)  
KEGG.out     = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "KEGG",        pval = pval_cutoff, qval = qval_cutoff)
HALLMARK.out = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "HALLMARK",    pval = pval_cutoff, qval = qval_cutoff) 
GO.out       = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "GO",          pval = pval_cutoff, qval = qval_cutoff) 
VAX.out      = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "VAX",         pval = pval_cutoff, qval = qval_cutoff)  
IMMUN.out    = geneEnrichments_v2(DEG.sig, bkgGenes = DEG$geneID, refDB = msigDB, enrichDB = "IMMUNESIGDB", pval = pval_cutoff, qval = qval_cutoff) 
```

### Summary of significant pathways 

```{r show enricher sum 2}
enricher.summary = data.frame(Type = c('Hallmark','KEGG','Reactome','GO','VAX','IMMUNESIGDB'),
                              Number_of_significant_sets = c(nrow(HALLMARK.out),nrow(KEGG.out),nrow(REACTOME.out),nrow(GO.out),nrow(VAX.out),nrow(IMMUN.out)))
create_dt(enricher.summary, paste0('Significant gene sets (p.adj < ', qval_cutoff, ")"))

```

### Hallmarks

```{r show enr hall 2}
tmp = HALLMARK.out
create_dt(tmp,'Significant hallmarks gene sets')
```

### KEGG

```{r show enr keg 2}
tmp = KEGG.out
create_dt(tmp,'Significant KEGG gene sets')
```

### Reactome

```{r show enr rea 2}
tmp = REACTOME.out
create_dt(tmp,'Significant reactome gene sets')
```

### GO BP

```{r show enr go 2,fig.height=10}
tmp = GO.out
create_dt(tmp,'Significant GO BP gene sets')
```

### VAX

```{r show enr vax 2,fig.height=10}
tmp = VAX.out
create_dt(tmp,'Significant VAX gene sets')
```

### IMMUNESIGDB

```{r show enr imm 2,fig.height=10}
tmp = IMMUN.out
create_dt(tmp,'Significant IMMUNESIGDB gene sets')
```

# Session Information

```{r save session}
save.image(file = '../data/session_3_ORA.rdata')
sessionInfo()
```